### PR TITLE
Fix `-inf` gradients in `normalize()` for float16 inputs (compute L2 norm in float32)

### LINKED ISSUE #23546: Bug: keras.src.ops.normalize() produces -inf gradients for small non-zero float16 inputs for the torch backend
### Description
`keras.src.ops.normalize()`'s L2 path performs its squared-norm computation, epsilon clamp, and `rsqrt` entirely in the input's original dtype. For `float16` inputs, this causes the gradient to become non-finite (`inf`/`NaN`) even for inputs that are not close to zero, because `epsilon**2` underflows to `0.0` in `float16` and `rsqrt`'s derivative overflows `float16`'s range at magnitudes well within normal usage. The example below shows a gradient of `-inf` for an input array filled with `0.01`, versus a small finite gradient in `float32`.

### Reproduction 
```
import os
os.environ["KERAS_BACKEND"] = "torch"

import numpy as np
import torch

from keras.src.ops import normalize

x_np = np.full((4, 2), 0.01, dtype=np.float16) 

x = torch.tensor(x_np, requires_grad=True)
y = normalize(x, axis=-1, order=2)
y.sum().backward()
grad = x.grad.detach().cpu().numpy()

print(f"gradient:\n{grad}")
# gradient:
# [[-inf -inf]
#  [-inf -inf]
#  [-inf -inf]
#  [-inf -inf]]

```

#### Output with float32
```
gradient:
[[7.6293945e-06 7.6293945e-06]
 [7.6293945e-06 7.6293945e-06]
 [7.6293945e-06 7.6293945e-06]
 [7.6293945e-06 7.6293945e-06]]
```

#### Expected Behavior
I would expect the gradient not to be -inf because it is far from that. It would be more logical for the gradient to be 0 everywhere. 

### Possible Fix
A possible fix would be to upcast the input to at least float32 before computing the squared norm, epsilon clamp, and rsqrt, and only casting back to the original dtype at the end. The following shows an example patch

```
diff --git a/keras/src/ops/nn.py b/keras/src/ops/nn.py
index c4d1b396a..ad6d7fab0 100644
--- a/keras/src/ops/nn.py
+++ b/keras/src/ops/nn.py
@@ -2722,14 +2722,23 @@ def _normalize(x, axis=-1, order=2, epsilon=None):
     if 2 == order:
         # A special case: L2 normalization with `x * rsqrt(...)`
         # instead of `x / sqrt(...)`. Clamp the squared norm before the
         # rsqrt so zero vectors get a finite gradient.
+        original_dtype = getattr(backend, 'standardize_dtype', lambda d: d)(x.dtype)
+        compute_dtype = getattr(backend, 'result_type', lambda a, b: 'float32')(x.dtype, 'float32')
+        x_cast = backend.cast(x, dtype=compute_dtype)
         square_sum = backend.numpy.sum(
-            backend.numpy.square(x), axis=axis, keepdims=True
+            backend.numpy.square(x_cast), axis=axis, keepdims=True
         )
+        # compute epsilon squared in compute dtype to avoid underflow
+        eps_sq = backend.cast(epsilon * epsilon, dtype=compute_dtype)
         inv_norm = backend.math.rsqrt(
-            backend.numpy.maximum(square_sum, epsilon * epsilon)
+            backend.numpy.maximum(square_sum, eps_sq)
         )
-        return x * inv_norm
+        out = x_cast * inv_norm
+        return backend.cast(out, dtype=original_dtype)
+
     norm = backend.linalg.norm(x, ord=order, axis=axis, keepdims=True)
     denom = backend.numpy.maximum(norm, epsilon)
     return backend.numpy.divide(x, denom)
```

### AI Disclosure
I used an LLM based approach for research to find this bug and I also used an agentic approach to find a plausible fix. I reviewed the found bug by checking if the behavior is still active on the most current version and agreeing that this behavior is indeed unintended. I also reviewed the proposed patch and take full responsibility for it.

### Environment
- Keras version: 78d44013f at the master branch
- Backend: NumPy (`KERAS_BACKEND=numpy`)
- Python version: Python 3.11.2
- OS: Debian GNU/Linux 12 (bookworm)

### PR DESCRIPTION
**Pull request by ManoharPaturi**
Body:

Fixes #23546

## Summary

`normalize()`'s L2 fast path computed the squared-norm, epsilon clamp, and
`rsqrt` in the **input dtype**. For float16, `epsilon**2` underflows to `0`
and the `rsqrt` derivative (`-0.5·s^-1.5`) overflows float16 range for small
squared-norms, so backward passes yield `-inf` gradients for inputs as small
as `0.01` — reproduced on the torch CPU backend with the issue's repro
before fixing.

The fix computes the L2 intermediates in float32 and casts the result back
to the original dtype, via `backend.result_type(x.dtype, "float32")` /
`backend.cast(...)` — the exact convention already used by
`_rms_normalization` and `_layer_normalization` in the same file. It is
placed in the core op (not the torch backend) so every backend benefits;
the non-L2 path is untouched, and the zero-vector gradient behavior from
#23075 is preserved.

## Repro (torch 2.14.0 CPU, before → after)

- before: gradient `[[-inf -inf] ...]`, `grad finite: False`
- after: gradient `[[-7.6e-06 ...]]` (float16 rounding of `-7.6293945e-06`),
  `grad finite: True`; output dtype still float16; float32 inputs unchanged

## Testing

- new regression test `test_normalize_l2_half_precision_gradients` in
  `keras/src/ops/nn_test.py`, modeled on the existing
  `test_normalize_l2_zero_vector_gradients`
- `KERAS_BACKEND=torch pytest keras/src/ops/nn_test.py -k normalize` → 7 passed
- `KERAS_BACKEND=numpy` → 5 passed, 2 skipped (gradient tests skip on numpy, per existing pattern)
- `KERAS_BACKEND=jax` → 7 passed
- full `nn_test.py` delta vs pristine master on this machine: exactly +1 passing test (pre-existing MPS float64 failures identical before/after)
- `ruff check` + `ruff format --check` clean on both changed files

## Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

## AI Assistance Disclosure

This pull request was prepared with AI coding assistance (an agent implemented
the fix and tests from issue #23546 under my direction). I have reviewed the
full diff, reproduced the bug and the fix locally on the torch CPU backend,
and verified the test results described below; I take responsibility for this
contribution and will drive it through review.

### CODE REVIEW

**Comment by gaga1313** [comment]
The changes LGTM! @hertschuh, if you can take a look that be great.

**Comment by gaga1313** [comment]
Hi @ManoharPaturi, thanks for the PR!

In `float16`, this overflow/underflow is expected due to the narrow dynamic range of half precision. Low-level `keras.ops` functions are expected to strictly respect the input `dtype` rather than silently upcasting to `float32`.

If we implicitly upcast `ops.normalize` to `float32` here, the same reasoning would apply to almost every other math/reduction op in `keras.ops`, which is not something we want to do at the low-level op layer. 

Instead, we recommend either:
1. Casting the input to `float32` before calling `ops.normalize(x)` (and casting back if needed).

**Comment by ManoharPaturi** [comment]
Thanks for the review and explanation @gaga1313! That makes total sense regarding keeping `keras.ops` strictly adhering to the input dtype without implicit upcasting. Appreciate the clarification!
